File: sum5frac.py
"""
For ints a, b, c, d, e.
Find unequal fractions 1/a, 1/b, 1/c, 1/d, 1/e such that their sum is exactly 1.
Only searches for solutions with a denominator under a specified maximum value.
Prints only unique solutions. For each solution, the fractions are summed in descending order.
"""

import sys
import argparse
parser = argparse.ArgumentParser(description="Finds 5 unique fractions that add to 1.")
parser.add_argument("-m", "--max", type=int, default=100, help="Maximum denominator to be used in fractions (default: 100)")
parser.add_argument("-i", "--iterations", type=int, default=-1, help="Print message after every i iterations (default: print nothing)")
args = parser.parse_args()

num_solutions = 0
max_denominator = args.max + 1
tolerance = .000000001
i = 0

def eprint(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)

for a in range(2, max_denominator):
    # Significant optimization.
    # Once a (the largest fraction in the 5 sum) becomes this large, there cannot be any more solutions
    # For example, when a >= 4, there will be no further solutions because 1/4 + 1/5 + 1/6 + 1/7 + 1/8 < 1
    if (1/a + 1/(a+1) + 1/(a+2) + 1/(a+3) + 1/(a+4)) < 1:
        break

    for b in range(a + 1, max_denominator):
        # Similar optimization as above
        if (1/b + 1/(b+1) + 1/(b+2) + 1/(b+3)) < 1 - (1/a):
            break

        for c in range(b + 1, max_denominator):
            if (1/c + 1/(c+1) + 1/(c+2)) < 1 - (1/a) - (1/b):
                break

            for d in range(c + 1, max_denominator):
                if (1/d + 1/(d+1)) < 1 - (1/a) - (1/b) - (1/c):
                    break

                for e in range(d + 1, max_denominator):
                    i += 1
                    if args.iterations != -1 and i % args.iterations == 0:
                        eprint(f"{i:,} iterations done.", flush=True)

                    if abs(1/a + 1/b + 1/c + 1/d + 1/e - 1) < tolerance:
                        print(f"1/{a} + 1/{b} + 1/{c} + 1/{d} + 1/{e}", flush=True)
                        num_solutions += 1

eprint("Number of solutions:", num_solutions)
eprint("Number of iterations:", i)

File: sum5frac.c
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>
#include <unistd.h>
#include <math.h>

#define OPTIONS "hm:i:"
#define TOLERANCE .0000001

int main(int argc, char **argv)
{
    int opt = 0;
    char *ptr;
    int max_denom = 100;
    int iterations = -1;

    while ((opt = getopt(argc, argv, OPTIONS)) != -1) {
        switch (opt) {
        case 'm':
            max_denom = strtoul(optarg, &ptr, 10);
            if (*ptr != '\0') {
                printf("Invalid max denominator\n");
                return 1;
            }
            break;
        case 'i':
            iterations = strtoul(optarg, &ptr, 10);
            if (*ptr != '\0') {
                printf("Invalid number of iterations\n");
                return 1;
            }
            break;
        case 'h':
	default:
            fprintf(stderr, "usage: ./sum5frac [-h] [-m MAX] [-i ITERATIONS]\n\n");
            fprintf(stderr, "Finds 5 unique fractions that add to 1.\n\n");
            fprintf(stderr, "optional arguments:\n");
            fprintf(stderr, "-h: show this help message and exit\n");
            fprintf(stderr, "-m MAX: Maximum denominator to be used in fractions (default: 100)\n");
            fprintf(stderr, "-i ITERATIONS: Print message after every i iterations (default: print nothing)\n");
            return 0;
        }
    }

    uint64_t i = 0;
    uint64_t num_solutions = 0;
    for (double a = 2; a <= max_denom; a++) {
        if (1.0/a + 1.0/(a+1) + 1.0/(a+2) + 1.0/(a+3) + 1.0/(a+4) < 1.0) {
            break;
        }
        for (double b = a + 1; b <= max_denom; b++) {
            if (1.0/b + 1.0/(b+1) + 1.0/(b+2) + 1.0/(b+3) < 1.0 - (1.0/a)) {
                break;
            }
            for (double c = b + 1; c <= max_denom; c++) {
                if (1.0/c + 1.0/(c+1) + 1.0/(c+2) < 1.0 - (1.0/a) - (1.0/b)) {
                    break;
                }
                for (double d = c + 1; d <= max_denom; d++) {
                    if (1.0/d + 1.0/(d+1) < 1.0 - (1.0/a) - (1.0/b) - (1.0/c)) {
                        break;
                    }
                    for (double e = d + 1; e <= max_denom; e++) {
                        i++;
                        if (iterations != -1 && i % iterations == 0) {
                            fprintf(stderr, "%" PRIu64 " iterations done.\n", i);
                        }

                        if (fabs(1.0/a + 1.0/b + 1.0/c + 1.0/d + 1.0/e - 1) < TOLERANCE) {
                            printf("1/%.0f + 1/%.0f + 1/%.0f + 1/%.0f + 1/%.0f\n", a, b, c, d, e);
                            num_solutions++;
                        }
                    }
                }
            }
        }
    }

    printf("Number of solutions: %" PRIu64 "\n", num_solutions);
    printf("Number of iterations: %" PRIu64 "\n", i);
    return 0;
}
